### ISSUE TITLE
#[tracing::instrument()] triggers infinite_loop

### ISSUE THREAD

**Issue by jayvdb** [rust-lang/rust-clippy#12291]
### Summary

The new `infinite_loop` clippy from https://github.com/rust-lang/rust-clippy/pull/11829 is being triggered for fns that use [`#[tracing::instrument()]`](https://docs.rs/tracing/latest/tracing/attr.instrument.html)

### Lint Name

infinite_loop

### Reproducer

I tried this code:

```rust
use actix_web::HttpResponse;

#[tracing::instrument()]
pub async fn my_route(
) -> HttpResponse {
    HttpResponse::InternalServerError().finish()
}
```

I saw this happen:

```
error: infinite loop detected
 --> src/routes/foo.rs:9:6
  |
9 | ) -> HttpResponse {
  |      ^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#infinite_loop
  = note: `-D clippy::infinite-loop` implied by `-D clippy::restriction`
  = help: to override `-D clippy::restriction` add `#[allow(clippy::infinite_loop)]`
help: if this is intentional, consider specifing `!` as function return
  |
7 |  -> !
  |

```

I expected to see this happen:

No error

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by Jarcho** [comment]
Needs either a call to `is_from_proc_macro` or `in_external_macro`.
